File: benchmarks/c_benchmark/dump_patt.c
/*
 *  dump_patt.c
 *  WebARKitLib-rs
 *
 *  This file is part of WebARKitLib-rs - WebARKit.
 *
 *  WebARKitLib-rs is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU Lesser General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  WebARKitLib-rs is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU Lesser General Public License for more details.
 *
 *  You should have received a copy of the GNU Lesser General Public License
 *  along with WebARKitLib-rs.  If not, see <http://www.gnu.org/licenses/>.
 *
 *  As a special exception, the copyright holders of this library give you
 *  permission to link this library with independent modules to produce an
 *  executable, regardless of the license terms of these independent modules,
 *  and to copy and distribute the resulting executable under terms of your
 *  choice, provided that you also meet, for each linked independent module,
 *  the terms and conditions of the license of that module. An independent
 *  module is a module which is neither derived from nor based on this
 *  library. If you modify this library, you may extend this exception to
 *  your version of the library, but you are not obligated to do so. If you
 *  do not wish to do so, delete this exception statement from your version.
 *
 *  Copyright 2026 WebARKit.
 *
 *  Author(s): Walter Perdan @kalwalt https://github.com/kalwalt
 *
 */

/*
 * Diagnostic harness for issue #103 (low CF from template matching).
 *
 * Runs `arDetectMarker` once on a luma raw image, then re-extracts the same
 * pattern patch via `arPattGetImage` using the contour data preserved in
 * `arHandle->markerInfo2[0]`. The extracted patch is dumped to disk as raw
 * bytes for byte-level comparison with the Rust port.
 *
 * NOTE: this is a sibling of `main.c`; it does NOT modify the timing
 * benchmark. Outputs:
 *   - <out_dir>/c_ext_patt.bin       raw ext_patt bytes (patt_size^2 * channels)
 *   - <out_dir>/c_ext_patt_meta.txt  patt_size, channels, mode, vertices, cf, id, dir
 */

#include <AR/ar.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

/* Compatibility helpers required by some WebARKitLib builds. */
char *cat(const char *file, size_t *bufSize_p) {
  FILE *fp;
  size_t len;
  char *s;
  if (!file)
    return NULL;
  fp = fopen(file, "rb");
  if (!fp)
    return NULL;
  fseek(fp, 0, SEEK_END);
  len = ftell(fp);
  fseek(fp, 0, SEEK_SET);
  if (!(s = (char *)malloc(len + 1))) {
    fclose(fp);
    return NULL;
  }
  if (fread(s, len, 1, fp) < 1 && len > 0) {
    free(s);
    fclose(fp);
    return NULL;
  }
  s[len] = '\0';
  fclose(fp);
  if (bufSize_p)
    *bufSize_p = len + 1;
  return s;
}
int test_d(const char *dir) { (void)dir; return 0; }
int mkdir_p(const char *path) { (void)path; return 0; }

static const char *pixfmt_name(AR_PIXEL_FORMAT f) {
  switch (f) {
  case AR_PIXEL_FORMAT_RGB:  return "AR_PIXEL_FORMAT_RGB";
  case AR_PIXEL_FORMAT_BGR:  return "AR_PIXEL_FORMAT_BGR";
  case AR_PIXEL_FORMAT_RGBA: return "AR_PIXEL_FORMAT_RGBA";
  case AR_PIXEL_FORMAT_BGRA: return "AR_PIXEL_FORMAT_BGRA";
  case AR_PIXEL_FORMAT_ABGR: return "AR_PIXEL_FORMAT_ABGR";
  case AR_PIXEL_FORMAT_ARGB: return "AR_PIXEL_FORMAT_ARGB";
  case AR_PIXEL_FORMAT_MONO: return "AR_PIXEL_FORMAT_MONO";
  default:                   return "<unknown>";
  }
}

static const char *mode_name(int m) {
  switch (m) {
  case AR_TEMPLATE_MATCHING_COLOR:                      return "AR_TEMPLATE_MATCHING_COLOR";
  case AR_TEMPLATE_MATCHING_MONO:                       return "AR_TEMPLATE_MATCHING_MONO";
  case AR_MATRIX_CODE_DETECTION:                        return "AR_MATRIX_CODE_DETECTION";
  case AR_TEMPLATE_MATCHING_COLOR_AND_MATRIX:           return "AR_TEMPLATE_MATCHING_COLOR_AND_MATRIX";
  case AR_TEMPLATE_MATCHING_MONO_AND_MATRIX:            return "AR_TEMPLATE_MATCHING_MONO_AND_MATRIX";
  default:                                              return "<unknown>";
  }
}

int main(int argc, char *argv[]) {
  if (argc < 5) {
    printf("Usage: %s <camera_para.dat> <patt.hiro> <luma.raw> <out_dir> [width] [height]\n",
           argv[0]);
    printf("  out_dir defaults expected: ../data\n");
    return 1;
  }

  const char *cparam_name = argv[1];
  const char *patt_name   = argv[2];
  const char *luma_name   = argv[3];
  const char *out_dir     = argv[4];
  int width  = (argc > 5) ? atoi(argv[5]) : 429;
  int height = (argc > 6) ? atoi(argv[6]) : 317;

  /* ---- ARParam + ARHandle setup, mirrors main.c ---- */
  ARParam cparam;
  if (arParamLoad(cparam_name, 1, &cparam) < 0) {
    fprintf(stderr, "ERROR: arParamLoad failed for %s\n", cparam_name);
    return 1;
  }
  arParamChangeSize(&cparam, width, height, &cparam);
  ARParamLT *cparamLT = arParamLTCreate(&cparam, AR_PARAM_LT_DEFAULT_OFFSET);

  ARHandle *arHandle = arCreateHandle(cparamLT);
  if (!arHandle) {
    fprintf(stderr, "ERROR: arCreateHandle failed\n");
    return 1;
  }
  arHandle->arPixelFormat = AR_PIXEL_FORMAT_MONO;
  /* arPatternDetectionMode is left at the default (AR_TEMPLATE_MATCHING_COLOR), which
     matches the simple Rust example; do not override. */

  ARPattHandle *pattHandle = arPattCreateHandle();
  if (!pattHandle || arPattLoad(pattHandle, patt_name) < 0) {
    fprintf(stderr, "ERROR: arPattLoad failed for %s\n", patt_name);
    return 1;
  }
  arPattAttach(arHandle, pattHandle);

  /* ---- Load raw luma image ---- */
  size_t data_size = (size_t)width * (size_t)height;
  unsigned char *luma_buffer = (unsigned char *)malloc(data_size);
  FILE *f = fopen(luma_name, "rb");
  if (!f || fread(luma_buffer, 1, data_size, f) != data_size) {
    fprintf(stderr, "ERROR: could not read %zu bytes from %s\n", data_size, luma_name);
    if (f) fclose(f);
    free(luma_buffer);
    return 1;
  }
  fclose(f);

  AR2VideoBufferT vbuf;
  memset(&vbuf, 0, sizeof(vbuf));
  vbuf.buff     = luma_buffer;
  vbuf.buffLuma = luma_buffer;
  vbuf.fillFlag = 1;

  /* ---- Run detection once ---- */
  if (arDetectMarker(arHandle, &vbuf) < 0) {
    fprintf(stderr, "ERROR: arDetectMarker failed\n");
    return 1;
  }

  int marker_num = arGetMarkerNum(arHandle);
  if (marker_num <= 0) {
    fprintf(stderr, "ERROR: no markers detected (marker_num=%d)\n", marker_num);
    return 1;
  }
  ARMarkerInfo *minfo = arGetMarker(arHandle);
  /* Pick the first marker with id >= 0; otherwise fall back to index 0. */
  int sel = 0;
  for (int i = 0; i < marker_num; i++) {
    if (minfo[i].id >= 0) { sel = i; break; }
  }

  /* The candidate's contour data lives in markerInfo2[sel] (assumes marker_info[i]
     corresponds to markerInfo2[i] for our single-detection case — true when no
     candidates were dropped between detection stages). */
  ARMarkerInfo2 *m2 = &arHandle->markerInfo2[sel];

  int patt_size      = pattHandle->pattSize;
  int patt_max_size  = patt_size * 2;
  int channels       = (arHandle->arPatternDetectionMode == AR_TEMPLATE_MATCHING_COLOR) ? 3 : 1;
  size_t ext_len     = (size_t)patt_size * (size_t)patt_size * (size_t)channels;
  unsigned char *ext_patt = (unsigned char *)malloc(ext_len);
  memset(ext_patt, 0, ext_len);

  /* Re-extract using the same contour data the matcher used internally.
     If the prototype differs in your AR/ar.h variant, adjust here. */
  int rv = arPattGetImage(arHandle->arImageProcMode, arHandle->arPatternDetectionMode,
                          patt_size, patt_max_size,
                          luma_buffer, width, height, arHandle->arPixelFormat,
                          m2->x_coord, m2->y_coord, m2->vertex,
                          arHandle->pattRatio, ext_patt);
  if (rv < 0) {
    fprintf(stderr, "ERROR: arPattGetImage returned %d\n", rv);
    free(ext_patt);
    return 1;
  }

  /* ---- Write outputs ---- */
  char path_bin[1024], path_meta[1024];
  snprintf(path_bin,  sizeof(path_bin),  "%s/c_ext_patt.bin",       out_dir);
  snprintf(path_meta, sizeof(path_meta), "%s/c_ext_patt_meta.txt", out_dir);

  FILE *fbin = fopen(path_bin, "wb");
  if (!fbin) {
    fprintf(stderr, "ERROR: could not open %s for writing\n", path_bin);
    return 1;
  }
  fwrite(ext_patt, 1, ext_len, fbin);
  fclose(fbin);

  FILE *fmeta = fopen(path_meta, "w");
  if (!fmeta) {
    fprintf(stderr, "ERROR: could not open %s for writing\n", path_meta);
    return 1;
  }
  fprintf(fmeta, "patt_size = %d\n",         patt_size);
  fprintf(fmeta, "channels  = %d\n",         channels);
  fprintf(fmeta, "mode      = %s\n",         mode_name(arHandle->arPatternDetectionMode));
  fprintf(fmeta, "pixfmt    = %s\n",         pixfmt_name(arHandle->arPixelFormat));
  fprintf(fmeta, "xsize     = %d\n",         width);
  fprintf(fmeta, "ysize     = %d\n",         height);
  fprintf(fmeta, "patt_ratio = %.6f\n",      arHandle->pattRatio);
  fprintf(fmeta, "selected  = %d (of %d markers)\n", sel, marker_num);
  /* The 4 contour vertices used by arPattGetImage. */
  for (int i = 0; i < 4; i++) {
    int idx = m2->vertex[i];
    fprintf(fmeta, "vertex[%d] = (%d, %d)\n", i, m2->x_coord[idx], m2->y_coord[idx]);
  }
  /* Ideal-space marker corners (post-distortion-correction). */
  for (int i = 0; i < 4; i++) {
    fprintf(fmeta, "marker_vertex[%d] = (%.4f, %.4f)\n", i, minfo[sel].vertex[i][0], minfo[sel].vertex[i][1]);
  }
  fprintf(fmeta, "id        = %d\n",         minfo[sel].id);
  fprintf(fmeta, "id_patt   = %d\n",         minfo[sel].idPatt);
  fprintf(fmeta, "cf        = %.6f\n",       minfo[sel].cf);
  fprintf(fmeta, "cf_patt   = %.6f\n",       minfo[sel].cfPatt);
  fprintf(fmeta, "dir       = %d\n",         minfo[sel].dir);
  fclose(fmeta);

  printf("Wrote %s (%zu bytes)\n", path_bin, ext_len);
  printf("Wrote %s\n", path_meta);
  printf("cf=%.4f id=%d dir=%d\n", minfo[sel].cf, minfo[sel].id, minfo[sel].dir);

  /* Cleanup */
  free(ext_patt);
  free(luma_buffer);
  arPattDetach(arHandle);
  arPattDeleteHandle(pattHandle);
  arDeleteHandle(arHandle);
  arParamLTFree(&cparamLT);

  return 0;
}

File: crates/core/examples/dump_patt.rs
/*
 *  dump_patt.rs
 *  WebARKitLib-rs
 *
 *  This file is part of WebARKitLib-rs - WebARKit.
 *
 *  WebARKitLib-rs is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU Lesser General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  WebARKitLib-rs is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU Lesser General Public License for more details.
 *
 *  You should have received a copy of the GNU Lesser General Public License
 *  along with WebARKitLib-rs.  If not, see <http://www.gnu.org/licenses/>.
 *
 *  As a special exception, the copyright holders of this library give you
 *  permission to link this library with independent modules to produce an
 *  executable, regardless of the license terms of these independent modules, and to
 *  copy and distribute the resulting executable under terms of your choice,
 *  provided that you also meet, for each linked independent module, the terms and
 *  conditions of the license of that module. An independent module is a module
 *  which is neither derived from nor based on this library. If you modify this
 *  library, you may extend this exception to your version of the library, but you
 *  are not obligated to do so. If you do not wish to do so, delete this exception
 *  statement from your version.
 *
 *  Copyright 2026 WebARKit.
 *
 *  Author(s): Walter Perdan @kalwalt https://github.com/kalwalt
 *
 */

//! Diagnostic harness for issue #103 (low CF from template matching).
//!
//! Mirrors the C-side `dump_patt` (`benchmarks/c_benchmark/dump_patt.c`):
//! runs `ar_detect_marker` once and re-extracts the matched pattern patch
//! via `ar_patt_get_image` using the contour data preserved in
//! `ar_handle.marker_info2[sel]`. The extracted patch is written to disk as
//! raw bytes for byte-level comparison with the C reference.
//!
//! Outputs (under `--out-dir`, default `benchmarks/data/`):
//!   - `rs_ext_patt.bin`        raw ext_patt bytes
//!   - `rs_ext_patt_meta.txt`   patt_size, channels, mode, vertices, cf, id, dir
//!
//! See [docs/issue-103-fix-plan.md](../../../../docs/issue-103-fix-plan.md)
//! and [benchmarks/data/README.md](../../../../benchmarks/data/README.md)
//! for the full diagnostic recipe.

use image::ImageReader;
use std::fs::File;
use std::io::Write;
use std::path::PathBuf;
use webarkitlib_rs::{
    arlog_e, arlog_i,
    image_proc::ARImageProcInfo,
    marker::ar_detect_marker,
    pattern::{ar_patt_get_image, ar_patt_load, AR_TEMPLATE_MATCHING_COLOR},
    types::{
        AR2VideoBufferT, AR2VideoTimestampT, ARHandle, ARMatrixCodeType, ARParam, ARParamLT,
        ARParamLTf, ARPixelFormat, AR_MATRIX_CODE_DETECTION,
    },
    version,
};

fn main() {
    webarkitlib_rs::arlog::ar_log_init_default();
    arlog_i!("WebARKitLib-rs v{}", version::get_version());
    arlog_i!("Diagnostic: dumping ar_patt_get_image output for issue #103");

    let args: Vec<String> = std::env::args().collect();
    let cparam_path = arg_or_default(&args, 1, "benchmarks/data/camera_para.dat");
    let patt_path = arg_or_default(&args, 2, "benchmarks/data/patt.hiro");
    let img_path = arg_or_default(&args, 3, "benchmarks/data/img.jpg");
    let out_dir = PathBuf::from(arg_or_default(&args, 4, "benchmarks/data"));

    arlog_i!("Loading camera parameters from {}", cparam_path);
    let param_file =
        File::open(&cparam_path).unwrap_or_else(|_| panic!("Failed to open {}", cparam_path));
    let param = ARParam::load(param_file).expect("Failed to read camera parameters");

    arlog_i!("Loading image {}", img_path);
    let img = ImageReader::open(&img_path)
        .unwrap_or_else(|_| panic!("Failed to open {}", img_path))
        .decode()
        .expect("Failed to decode image");
    let width = img.width() as i32;
    let height = img.height() as i32;
    let luma_img = img.to_luma8();

    // Export the raw luma so the C dumper consumes byte-identical input.
    let hiro_raw_path = out_dir.join("hiro.raw");
    {
        let mut f = File::create(&hiro_raw_path)
            .unwrap_or_else(|_| panic!("Failed to create {}", hiro_raw_path.display()));
        f.write_all(luma_img.as_raw()).expect("write raw luma");
    }
    arlog_i!(
        "Exported raw luma to {} ({} bytes)",
        hiro_raw_path.display(),
        width * height
    );

    // Otsu threshold (same arithmetic the simple example uses).
    let mut ipi = ARImageProcInfo::new(width, height);
    let otsu_thresh = ipi
        .luma_hist_and_otsu(luma_img.as_raw())
        .expect("Otsu failed");
    arlog_i!("Otsu threshold: {}", otsu_thresh);

    // Identity LT (skip lens distortion correction for the diagnostic).
    let mut param_ltf = ARParamLTf::default();
    param_ltf.xsize = width;
    param_ltf.ysize = height;
    param_ltf.o2i = vec![0.0; (width * height * 2) as usize];
    for y in 0..height {
        for x in 0..width {
            let idx = ((y * width + x) * 2) as usize;
            param_ltf.o2i[idx] = x as f32;
            param_ltf.o2i[idx + 1] = y as f32;
        }
    }
    let mut param_lt = Box::new(ARParamLT { param, param_ltf });

    let mut ar_handle = ARHandle::default();
    ar_handle.xsize = width;
    ar_handle.ysize = height;
    ar_handle.set_pixel_format(ARPixelFormat::MONO);
    ar_handle.ar_image_proc_mode = 0; // FrameImage
    ar_handle.ar_pattern_detection_mode = AR_TEMPLATE_MATCHING_COLOR;
    ar_handle.ar_labeling_mode = 0;
    ar_handle.ar_labeling_thresh = otsu_thresh as i32;
    ar_handle.patt_ratio = 0.5;
    ar_handle.matrix_code_type = ARMatrixCodeType::Code3x3;
    ar_handle.ar_param_lt = &mut *param_lt;

    let mut patt_handle = webarkitlib_rs::types::ARPattHandle::default();
    patt_handle.patt_num_max = 50;
    patt_handle.patt_size = 16;
    patt_handle.pattf = vec![0; 50];
    patt_handle.patt = vec![vec![0; 16 * 16 * 3 * 4]; 50];
    patt_handle.pattpow = vec![0.0; 50 * 4];
    patt_handle.patt_bw = vec![vec![0; 16 * 16 * 4]; 50];
    patt_handle.pattpow_bw = vec![0.0; 50 * 4];

    arlog_i!("Loading hiro pattern from {}", patt_path);
    if let Err(e) = ar_patt_load(&mut patt_handle, &patt_path) {
        arlog_e!("ar_patt_load failed: {}", e);
        return;
    }

    let mut boxed_patt_handle = Box::new(patt_handle);
    ar_handle.patt_handle = &mut *boxed_patt_handle;

    let luma_buffer_vec = luma_img.into_raw();
    let frame = AR2VideoBufferT {
        buff: Some(luma_buffer_vec.clone()),
        buff_luma: Some(luma_buffer_vec.clone()),
        buf_planes: None,
        buf_plane_count: 0,
        fill_flag: true,
        time: AR2VideoTimestampT { sec: 1, usec: 0 },
    };

    arlog_i!("Running ar_detect_marker...");
    if let Err(e) = ar_detect_marker(&mut ar_handle, &frame) {
        arlog_e!("ar_detect_marker failed: {}", e);
        return;
    }
    if ar_handle.marker_num <= 0 {
        arlog_e!("No markers detected (marker_num={})", ar_handle.marker_num);
        return;
    }

    // Pick the first marker (matches the C dumper's selection rule).
    let sel: usize = (0..ar_handle.marker_num as usize)
        .find(|&i| ar_handle.marker_info[i].id >= 0)
        .unwrap_or(0);
    let marker = ar_handle.marker_info[sel].clone();
    let info2 = ar_handle.marker_info2[sel].clone();

    let patt_size = boxed_patt_handle.patt_size as usize;
    let patt_max_size = patt_size * 2;
    let channels = if ar_handle.ar_pattern_detection_mode == AR_TEMPLATE_MATCHING_COLOR {
        3
    } else {
        1
    };
    let ext_len = patt_size * patt_size * channels;
    let mut ext_patt = vec![0u8; ext_len];

    let vertex_indices: [usize; 4] = [
        info2.vertex[0] as usize,
        info2.vertex[1] as usize,
        info2.vertex[2] as usize,
        info2.vertex[3] as usize,
    ];

    arlog_i!("Calling ar_patt_get_image to re-extract the patch...");
    if let Err(e) = ar_patt_get_image(
        ar_handle.ar_image_proc_mode,
        ar_handle.ar_pattern_detection_mode,
        patt_size,
        patt_max_size,
        &luma_buffer_vec,
        width as usize,
        height as usize,
        ar_handle.ar_pixel_format,
        &info2.x_coord,
        &info2.y_coord,
        &vertex_indices,
        ar_handle.patt_ratio,
        &mut ext_patt,
    ) {
        arlog_e!("ar_patt_get_image failed: {}", e);
        return;
    }

    let bin_path = out_dir.join("rs_ext_patt.bin");
    let meta_path = out_dir.join("rs_ext_patt_meta.txt");

    File::create(&bin_path)
        .and_then(|mut f| f.write_all(&ext_patt))
        .expect("write rs_ext_patt.bin");

    let mut meta = String::new();
    meta.push_str(&format!("patt_size = {}\n", patt_size));
    meta.push_str(&format!("channels  = {}\n", channels));
    meta.push_str(&format!(
        "mode      = {}\n",
        mode_name(ar_handle.ar_pattern_detection_mode)
    ));
    meta.push_str(&format!(
        "pixfmt    = {}\n",
        pixfmt_name(ar_handle.ar_pixel_format)
    ));
    meta.push_str(&format!("xsize     = {}\n", width));
    meta.push_str(&format!("ysize     = {}\n", height));
    meta.push_str(&format!("patt_ratio = {:.6}\n", ar_handle.patt_ratio));
    meta.push_str(&format!(
        "selected  = {} (of {} markers)\n",
        sel, ar_handle.marker_num
    ));
    for i in 0..4 {
        let idx = vertex_indices[i];
        meta.push_str(&format!(
            "vertex[{}] = ({}, {})\n",
            i, info2.x_coord[idx], info2.y_coord[idx]
        ));
    }
    for i in 0..4 {
        meta.push_str(&format!(
            "marker_vertex[{}] = ({:.4}, {:.4})\n",
            i, marker.vertex[i][0], marker.vertex[i][1]
        ));
    }
    meta.push_str(&format!("id        = {}\n", marker.id));
    meta.push_str(&format!("id_patt   = {}\n", marker.id_patt));
    meta.push_str(&format!("cf        = {:.6}\n", marker.cf));
    meta.push_str(&format!("cf_patt   = {:.6}\n", marker.cf_patt));
    meta.push_str(&format!("dir       = {}\n", marker.dir));
    File::create(&meta_path)
        .and_then(|mut f| f.write_all(meta.as_bytes()))
        .expect("write rs_ext_patt_meta.txt");

    arlog_i!("Wrote {} ({} bytes)", bin_path.display(), ext_patt.len());
    arlog_i!("Wrote {}", meta_path.display());
    arlog_i!("cf={:.4} id={} dir={}", marker.cf, marker.id, marker.dir);
}

fn arg_or_default(args: &[String], idx: usize, default: &str) -> String {
    args.get(idx)
        .cloned()
        .unwrap_or_else(|| default.to_string())
}

fn mode_name(m: i32) -> &'static str {
    if m == AR_TEMPLATE_MATCHING_COLOR {
        "AR_TEMPLATE_MATCHING_COLOR"
    } else if m == webarkitlib_rs::pattern::AR_TEMPLATE_MATCHING_MONO {
        "AR_TEMPLATE_MATCHING_MONO"
    } else if m == AR_MATRIX_CODE_DETECTION {
        "AR_MATRIX_CODE_DETECTION"
    } else {
        "<unknown>"
    }
}

fn pixfmt_name(f: ARPixelFormat) -> &'static str {
    match f {
        ARPixelFormat::RGB => "AR_PIXEL_FORMAT_RGB",
        ARPixelFormat::BGR => "AR_PIXEL_FORMAT_BGR",
        ARPixelFormat::RGBA => "AR_PIXEL_FORMAT_RGBA",
        ARPixelFormat::BGRA => "AR_PIXEL_FORMAT_BGRA",
        ARPixelFormat::ABGR => "AR_PIXEL_FORMAT_ABGR",
        ARPixelFormat::ARGB => "AR_PIXEL_FORMAT_ARGB",
        ARPixelFormat::MONO => "AR_PIXEL_FORMAT_MONO",
        _ => "<unknown>",
    }
}
